Libraries/plist/Tools/PlistBuddy.cpp

```diff
@@ -0,0 +1,546 @@
+#undef INDENT
+
+    return (error.empty() ? 0 : -1);
+}
+
+static void
+CommandHelp()
+{
+#define INDENT "  "
+    fprintf(stderr, "\nCommands help:\n");
+    fprintf(stderr, INDENT "Help - Print this information\n");
+    fprintf(stderr, INDENT "Exit - Exits this program\n");
+    fprintf(stderr, INDENT "Print [<KeyPath>] - Print value at KeyPath. (default KeyPath = root)\n");
+    fprintf(stderr, INDENT "Save - Save the changed plist file\n");
+    fprintf(stderr, INDENT "Set <KeyPath> <Value> - Set value at KeyPath to Value\n");
+    fprintf(stderr, INDENT "Add <KeyPath> <Type> <Value> - Set value at KeyPath to Value\n");
+    fprintf(stderr, INDENT "Clear <Type> - Clears all data, and sets root to of the given type\n");
+    fprintf(stderr, INDENT "Delete <KeyPath> - Removes entry at KeyPath\n");
+    fprintf(stderr, "\n<KeyPath>\n");
+    fprintf(stderr, INDENT ":= \"\"                             => root object\n");
+    fprintf(stderr, INDENT ":= <KeyPath>[:<Dictionary Key>]   => indexes into dictionary\n");
+    fprintf(stderr, INDENT ":= <KeyPath>[:<Array Index>]      => indexes into Array\n");
+    fprintf(stderr, "\n<Type> := (string|dictionary|array|bool|real|integer|date|data)\n\n");
+#undef INDENT
+}
+
+static void
+ParseCommandKeyPathString(std::string const &keyPathString, std::queue<std::string> *keyPath) {
+    auto loc = keyPathString.find(':');
+    if (loc == std::string::npos) {
+        if (keyPathString.size() > 0) {
+            keyPath->push(keyPathString);
+        }
+    } else {
+        if (loc > 0) {
+            keyPath->push(keyPathString.substr(0, loc));
+        }
+        ParseCommandKeyPathString(keyPathString.substr(loc + 1), keyPath);
+    }
+}
+
+static std::string parseCommandValueString(std::vector<std::string>::const_iterator begin, std::vector<std::string>::const_iterator end) {
+    if (begin == end) {
+        return "";
+    }
+    std::stringstream sstream;
+    for (auto it = begin; it != end; it++) {
+        if (it != begin) {
+            sstream << ' ';
+        }
+        sstream << it->c_str();
+    }
+
+    std::string valueString = sstream.str();
+    if (valueString.front() == '\"' && valueString.back() == '\"') {
+        // strip double quote if necessary
+        valueString = valueString.substr(1, valueString.size() - 2);
+    }
+    return valueString;
+}
+
+static plist::Object *getObjectAtKeyPath(plist::Object *object, std::queue<std::string> &remainingKeys, bool leaveLastKey = true) {
+    if (remainingKeys.empty() || (leaveLastKey && remainingKeys.size() == 1)) {
+        return object;
+    } else if (!object) {
+        fprintf(stderr, "Invalid key path (indexing into null object)\n");
+        return nullptr;
+    }
+
+    std::string currentKey = remainingKeys.front();
+
+    plist::Object *subObject = nullptr;
+    switch (object->type()) {
+        case plist::ObjectType::Dictionary:
+            subObject = plist::CastTo<plist::Dictionary>(object)->value(currentKey);
```

Comment:
Rather than the `switch`, you could do this:

    if (plist::Dictionary *dictionary = plist::CastTo<plist::Dictionary>(object)) {
        subObject = dictionary->value(currentKey);
    } else if (plist::Array *array = plist::CastTo<plist::Array>(object)) {
        /* ... */
    } else {
        /* ... */
    }

But not too different either way...